dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/tei/query/QueryContext.java

```diff
@@ -0,0 +1,258 @@
+    }
+
+    private Pair<Renderable, Renderable> toOrderSubQueryWithCondition( AnalyticsSortingParams analyticsSortingParams )
+    {
+        return LeftJoinQueryBuilder.of( analyticsSortingParams, this );
+    }
+
+    private List<AnalyticsSortingParams> getOrdersForSubQuery( TeiQueryParams teiQueryParams )
+    {
+        return teiQueryParams.getCommonParams().getOrderParams().stream()
+            .filter( this::needsSubQuery )
+            .collect( Collectors.toList() );
+    }
+
+    private boolean needsSubQuery( AnalyticsSortingParams analyticsSortingParams )
+    {
+        DimensionIdentifier<Program, ProgramStage, DimensionParam> orderBy = analyticsSortingParams.getOrderBy();
+        return orderBy.hasProgram();
+    }
+
+    private Where getWhere()
+    {
+        // conditions on programs (is enrolled in program)
+        Stream<Renderable> programConditions = getTeiQueryParams()
+            .getCommonParams()
+            .getPrograms()
+            .stream()
+            .map( program -> () -> TEI_ALIAS + "." + program.getUid() );
+
+        // conditions on filters/dimensions
+        Stream<Renderable> dimensionConditions = getTeiQueryParams().getCommonParams().getDimensionIdentifiers()
+            .stream()
+            .map( this::toConditions )
+            .map( OrCondition::of );
+
+        return Where.of(
+            AndCondition.of(
+                Stream.concat(
+                    programConditions,
+                    dimensionConditions )
+                    .collect( Collectors.toList() ) ) );
+    }
+
+    private List<Renderable> toConditions(
+        List<DimensionIdentifier<Program, ProgramStage, DimensionParam>> dimensionIdentifiers )
+    {
+        return dimensionIdentifiers.stream()
+            .filter( di -> di.getDimension().hasRestrictions() )
+            .collect( Collectors.groupingBy( di -> di.getDimension().getDimensionParamObjectType() ) )
+            .entrySet().stream()
+            .map( entry -> toCondition( entry.getKey(), entry.getValue() ) )
+            .collect( Collectors.toList() );
+    }
+
+    private Renderable toCondition( DimensionParamObjectType type,
+        List<DimensionIdentifier<Program, ProgramStage, DimensionParam>> dimensionIdentifiers )
+    {
+        // TODO: depending on the type, we should build a proper condition
+        if ( type == DimensionParamObjectType.DATA_ELEMENT )
+        {
+            return AndCondition.of(
+                dimensionIdentifiers.stream()
+                    .map( dimensionIdentifier -> EventDataValueCondition.of( dimensionIdentifier, this ) )
+                    .collect( Collectors.toList() ) );
+        }
+        return () -> StringUtils.EMPTY;
+    }
+
+    private static class ParameterManager
+    {
+        private int parameterIndex = 0;
+
+        @Getter
+        private final Map<Integer, Object> parametersByPlaceHolder = new HashMap<>();
+
+        public String bindParamAndGetIndex( Object param )
```

Comment:
Method is not in use, parameterIndex can be removed as well

Reply:
it is used by upper class, through "@Delegate" annotation. Intellij doesn't get aware of it